core/src/apps/common/paths.py

```diff
@@ -18,17 +28,219 @@
+
+            elif component == "coin_type":
+                # substitute SLIP-44 ids
+                self.schema.append(set(parse(s) for s in slip44_id))
+
+            else:
+                # plain constant
+                self.schema.append((parse(component),))
+
+    def match(self, path: Bip32Path) -> bool:
+        # The path must not be _shorter_ than schema. It may be longer.
+        if len(path) < len(self.schema):
+            return False
+
+        path_iter = iter(path)
+        # iterate over length of schema, consuming path components
+        for expected in self.schema:
+            value = next(path_iter)
+            if value not in expected:
+                return False
+
+        # iterate over remaining path components
+        for value in path_iter:
+            if value not in self.trailing_components:
+                return False
+
+        return True
+
+    if __debug__:
+
+        def __repr__(self) -> str:
+            components = ["m"]
+
+            def unharden(item: int) -> int:
+                return item ^ (item & HARDENED)
+
+            for component in self.schema:
+                if isinstance(component, range):
+                    a, b = component.start, component.stop - 1
+                    components.append(
+                        "[{}-{}]{}".format(
+                            unharden(a), unharden(b), "'" if a & HARDENED else ""
+                        )
+                    )
+                else:
+                    component_str = ",".join(str(unharden(i)) for i in component)
+                    if len(component) > 1:
+                        component_str = "[" + component_str + "]"
+                    if next(iter(component)) & HARDENED:
+                        component_str += "'"
+                    components.append(component_str)
+
+            if self.trailing_components:
+                for key, val in self.WILDCARD_RANGES.items():
+                    if self.trailing_components is val:
+                        components.append(key)
+                        break
+                else:
+                    components.append("???")
+
+            return "<schema:" + "/".join(components) + ">"
+
+
+class _AlwaysMatchingSchema:
+    @staticmethod
+    def match(path: Bip32Path) -> bool:
+        return True
+
+
+class _NeverMatchingSchema:
+    @staticmethod
+    def match(path: Bip32Path) -> bool:
+        return False
+
+
+# type objects _AlwaysMatchingSchema and _NeverMatching schema conform to the
+# PathSchemaType protocol, but mypy fails to recognize this due to:
+# https://github.com/python/mypy/issues/4536,
+# hence the following trickery
+AlwaysMatchingSchema: PathSchemaType = _AlwaysMatchingSchema  # type: ignore
+NeverMatchingSchema: PathSchemaType = _NeverMatchingSchema  # type: ignore
+
+PATTERN_BIP44 = "m/44'/coin_type'/account'/change/address_index"
+PATTERN_BIP44_PUBKEY = "m/44'/coin_type'/account'/*"
+PATTERN_SEP5 = "m/44'/coin_type'/account'"
```

Comment:
Can you please include comment with a link to the spec? It's quite hard to search for:/

Reply:
b90dfb73037c09b990ad4738f037339a57612ef7